[flink] integrate DELETE filter pushdown for flink-1.17

legendtkl opened the pull request:
<!-- Please specify the module before the PR name: [core] ... or [flink] ... -->

### Purpose
Linked issue: close https://github.com/apache/incubator-paimon/issues/989

### Brief change
Supported filters:
    1. where primary key = x
    2. where primary key = x or key = y
    3. where primary key in (x, y, z)
    4. where partition key = x

Note: primary key between filter now is not supported in Paimon, ignore it now and let's start a new ticket for it later.

### Tests
UT:
* ReadWriteTableITCase#testDeletePushDownWithPrimaryKey
* ReadWriteTableITCase#testDeletePushDownWithPartitionKey


### Documentation

<!-- Does this change introduce a new feature -->

JingsongLi commented on paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/FlinkTableSink.java:
```diff
@@ -112,17 +132,95 @@ public RowLevelUpdateInfo applyRowLevelUpdate(
     @Override
     public RowLevelDeleteInfo applyRowLevelDelete(
             @Nullable RowLevelModificationScanContext rowLevelModificationScanContext) {
+        if (supportDelete()) {
```
`void checkDeletable`?

JingsongLi commented on paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/FlinkTableSink.java:
```diff
@@ -112,17 +132,95 @@ public RowLevelUpdateInfo applyRowLevelUpdate(
+                return false;
+            }
+
+            predicates = new ArrayList<>();
```
local variable, don't assign to class field when return false.

JingsongLi commented on paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/FlinkTableSink.java:
```diff
@@ -135,4 +233,50 @@ public RowLevelDeleteInfo applyRowLevelDelete(
         }
     }
+
+    private boolean shouldPushdownDeleteFilter(Predicate predicate, int predicateSize) {
```
`canPushdownDeleteFilter`

JingsongLi commented on paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/FlinkTableSink.java:
```diff
@@ -135,4 +233,50 @@ public RowLevelDeleteInfo applyRowLevelDelete(
         }
     }
+
+    private boolean shouldPushdownDeleteFilter(Predicate predicate, int predicateSize) {
```
Use `FunctionVisitor` to visit predicate?

legendtkl replied:
hi, @JingsongLi , is the visitor too complicated for this to implement all the method visitXXX of interface FunctionVisitor?

When the filter is `where id=1 or id=2 or id=3` or `where id in (1,2,3)`, it will be converted to 
```sql
               compound(OR)
                   /     \
         compound(OR)     leaf(Equal, id, 1)
         /        \
leaf(Equal,id,2)   leaf(Equal,id,3)
```

We just need to check all the compoundFilter's func is OR and all the leafFilter's func is Equal and field is in primary keys.

JingsongLi replied:
We can just throw exception if encounter an unsupported predicate.
And catch the specific exception outer.

This is a common code style in flink-calcite.

JingsongLi replied:
And we can support `IN` too.

klion26 commented:
Hi, can this pr fix the problem `Could not execute SQL statement. Reason:
org.apache.flink.table.api.TableException: DELETE statement is not supported for streaming mode now.`?

legendtkl commented:
> Hi, can this pr fix the problem `Could not execute SQL statement. Reason: org.apache.flink.table.api.TableException: DELETE statement is not supported for streaming mode now.`?

@klion26 , no, the exception is threw by Flink TableEnv.

yuzelin commented on paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/FlinkTableSink.java:
```diff
@@ -112,17 +132,95 @@ public RowLevelUpdateInfo applyRowLevelUpdate(
+            }
+
+            LeafPredicate leaf = (LeafPredicate) predicates.get(0);
+            String[] args =
+                    new String[] {
+                        "--path",
+                        tablePath,
+                        "--partition",
+                        String.format("%s=%s", leaf.fieldName(), leaf.literals().get(0))
+                    };
+
+            Optional<Action> action = DropPartitionAction.create(args);
+            try {
+                action.get().run();
+            } catch (Exception e) {
+                throw new RuntimeException(
+                        String.format("DropPartitionAction %s failed.", Arrays.toString(args)), e);
+            }
```
I think you can directly use `commit.dropPartitions` instead of building an action.

legendtkl replied:
hi, @yuzelin , do you mean 
```code
action = DropPartitionAction(xx)
action.run()
```

But in this case, we need to parse the parameters for DropPartitionAction construct method. This might cause some duplicated code and some complicated cases.

yuzelin replied:
I mean you can:
1. get `FileStoreCommit` from table;
2. call `FileStoreCommit#dropPartitions` directly.

legendtkl replied:
> I mean you can:
> 
> 1. get `FileStoreCommit` from table;
> 2. call `FileStoreCommit#dropPartitions` directly.

@yuzelin  got it. Resolved.

yuzelin commented on paimon-core/src/main/java/org/apache/paimon/table/TableUtils.java:
```diff
@@ -40,10 +40,11 @@ public class TableUtils {
      * <p>NOTE: This method is only suitable for deletion of small amount of data.
      */
-    public static void deleteWhere(Table table, List<Predicate> filters) {
+    public static long deleteWhere(Table table, List<Predicate> filters) {
```
Add `@return`.

yuzelin reviewed (commented):
Thanks for contributing, left some comments.

legendtkl commented:
Thanks for the review, @JingsongLi @yuzelin.

I have addressed the comment and leave some comments from my view.

Would you help review it again when you're available? thanks.

yuzelin reviewed (approved):
+1

JingsongLi reviewed (commented):
Regarding the implementation of `DeletePushDownFunctionVisitor`, I feel that it is OK to return bool. For `AND`, as long as there is a true, it can return true; for `OR`, as long as there is a false, it returns false.

JingsongLi reviewed (changes requested):
-1, we can improve more.

JingsongLi commented on paimon-common/src/main/java/org/apache/paimon/predicate/DeletePushDownFunctionVisitor.java:
```diff
@@ -0,0 +1,128 @@
+    @Override
+    public Boolean visitEqual(FieldRef fieldRef, Object literal) {
+        if (partitionKeys.contains(fieldRef.name())) {
+            if (predicatesSize == 1) {
```
Why?

legendtkl replied:
@JingsongLi  partition key filter could be push down when there is only one filter, such as `when partition key = xx`. When the partition key filter is mixed with others, do not push down.

JingsongLi commented on paimon-common/src/main/java/org/apache/paimon/predicate/DeletePushDownFunctionVisitor.java:
```diff
@@ -0,0 +1,128 @@
+
+    @Override
+    public Boolean visitAnd(List<Boolean> children) {
+        if (children.size() != 2) {
```
Why check 2?

legendtkl replied:
follow legacy code, can remove it

JingsongLi commented on paimon-common/src/main/java/org/apache/paimon/predicate/DeletePushDownFunctionVisitor.java:
```diff
@@ -0,0 +1,128 @@
+
+    @Override
+    public Boolean visitOr(List<Boolean> children) {
+        if (children.size() != 2) {
```
Why check 2?

legendtkl replied:
follow legacy code, can remove it

JingsongLi commented:
I feel like there are still some issues:
- For the second visitor, what you need to determine is whether all pk (pk including partition keys) equivalence filters are included, others filters can be ignore.
- Does ITCase seem to be of little use for testing? What do you need is UT to verify if it has been pushed down?

JingsongLi reviewed (approved):
+1
